src/aggregator/aggregator/generic_elem.go

```diff
@@ -358,40 +364,47 @@ func (e *GenericElem) expireFlushState() {
 	}
 }
 
-// return the timestamp in the values map that is before the provided time. returns false if the provided time is the
-// smallest time or the map is empty.
-func (e *GenericElem) previousStartAlignedWithLock(timestamp xtime.UnixNano) (xtime.UnixNano, bool) {
+// return the previous aggregation before the provided time. returns false if the provided time is the
+// earliest time or the map is empty.
+func (e *GenericElem) prevAggWithLock(agg timedAggregation) (timedAggregation, bool) {
 	if len(e.values) == 0 {
-		return 0, false
+		return timedAggregation{}, false
+	}
+	if agg.prevStart != 0 {
+		prevAgg, ok := e.values[agg.prevStart]
+		return prevAgg, ok
 	}
+
 	resolution := e.sp.Resolution().Window
-	// ensure the input is start aligned and then calculate the previous start time.
-	startAligned := timestamp.Truncate(resolution).Add(-resolution)
-	for !startAligned.Before(e.minStartTime) {
-		_, ok := e.values[startAligned]
+	startTime := agg.startAt.Add(-resolution)
+	for !startTime.Before(e.minStartTime) {
+		agg, ok := e.values[startTime]
 		if ok {
-			return startAligned, true
+			return agg, true
 		}
-		startAligned = startAligned.Add(-resolution)
+		startTime = startTime.Add(-resolution)
 	}
-	return 0, false
+	return timedAggregation{}, false
 }
 
 // return the next aggregation after the provided time. returns false if the provided time is the
 // largest time or the map is empty.
-func (e *GenericElem) nextAggWithLock(startAligned xtime.UnixNano, targetNanos int64,
-	isEarlierThanFn isEarlierThanFn) (timedAggregation, bool) {
+func (e *GenericElem) nextAggWithLock(agg timedAggregation) (timedAggregation, bool) {
 	if len(e.values) == 0 {
 		return timedAggregation{}, false
 	}
+	if agg.nextStart != 0 {
+		nextAgg, ok := e.values[agg.nextStart]
+		return nextAgg, ok
+	}
 	resolution := e.sp.Resolution().Window
-	ts := startAligned.Add(resolution)
-	for isEarlierThanFn(int64(ts), resolution, targetNanos) {
```

Comment:
Is it a problem we're not using that custom `isEarlierFn`?

Reply:
no, because are no longer needing to compare against `targetNanos`, since we have a max map entry.